File: OnlyCue/MIDI/MTCFrame.swift
import Foundation

/// The MIDI Timecode wire format — pure byte encoding, no CoreMIDI (epic #794).
///
/// Two message shapes carry MTC:
///
/// - **Quarter-frame** (`F1 <data>`) — the running stream. Each message carries
///   one nibble of the timecode, so a complete value takes **eight** messages
///   and therefore spans **two** frames. Emitted continuously while the
///   transport runs.
/// - **Full Frame** (`F0 7F 7F 01 01 hh mm ss ff F7`) — a Universal Real Time
///   SysEx that jams a receiver to an absolute position in one message. Sent on
///   locate (play-start, seek, and scrubbing while paused) so a console does not
///   have to wait two frames to know where it is.
///
/// Mirrors `MIDIMessage` in spirit: pure, hardware-free, and the place where all
/// the bit-twiddling is pinned by tests. `MTCOutput` only pushes these bytes at
/// the right times.
///
/// **Rate bits are two bits wide** — MTC encodes exactly 24 / 25 / 29.97-DF / 30
/// and has no 30-fps-drop-frame code. `SMPTEFramerate.fps30drop` therefore takes
/// the 29.97-DF slot while still being clocked at 30 fps, which is the same
/// simplification `LTCEncoder` already ships (ADR-019, reaffirmed by ADR-032).
/// The invariant that matters is that LTC and MTC never disagree.
enum MTCFrame {

    /// Status byte introducing a quarter-frame message.
    static let quarterFrameStatus: UInt8 = 0xF1

    /// Quarter-frame messages needed to transmit one complete timecode value.
    static let piecesPerTimecode = 8

    /// Frames spanned by one full eight-message quarter-frame sequence.
    static let framesPerSequence = 2

    // MARK: - Rate bits

    /// The two-bit MTC rate code for `rate`.
    static func rateBits(for rate: SMPTEFramerate) -> UInt8 {
        switch rate {
        case .fps24:     return 0b00
        case .fps25:     return 0b01
        case .fps30drop: return 0b10   // 29.97 drop — the only drop-frame slot MTC has
        case .fps30:     return 0b11
        }
    }

    // MARK: - Quarter frames

    /// The data byte for quarter-frame `piece` (0...7) of `timecode`.
    ///
    /// Layout is `0nnn dddd`: `nnn` is the piece index, `dddd` the payload
    /// nibble. Pieces run least-significant-first — frames, then seconds, then
    /// minutes, then hours — with the rate bits riding in piece 7 alongside the
    /// single high bit of the hour.
    ///
    /// `piece` is clamped to 0...7: an out-of-range index is a caller bug, and
    /// clamping keeps a malformed byte (one whose high nibble names a different
    /// piece) off the wire.
    static func quarterFrameByte(piece: Int, timecode: Timecode) -> UInt8 {
        let index = min(max(piece, 0), piecesPerTimecode - 1)
        let payload: UInt8
        switch index {
        case 0: payload = UInt8(timecode.frames) & 0x0F
        case 1: payload = (UInt8(timecode.frames) >> 4) & 0x01
        case 2: payload = UInt8(timecode.seconds) & 0x0F
        case 3: payload = (UInt8(timecode.seconds) >> 4) & 0x03
        case 4: payload = UInt8(timecode.minutes) & 0x0F
        case 5: payload = (UInt8(timecode.minutes) >> 4) & 0x03
        case 6: payload = UInt8(timecode.hours) & 0x0F
        default: payload = (rateBits(for: timecode.rate) << 1) | ((UInt8(timecode.hours) >> 4) & 0x01)
        }
        return (UInt8(index) << 4) | payload
    }

    /// The complete two-byte quarter-frame message for `piece` of `timecode`.
    static func quarterFrameMessage(piece: Int, timecode: Timecode) -> [UInt8] {
        [quarterFrameStatus, quarterFrameByte(piece: piece, timecode: timecode)]
    }

    // MARK: - Full Frame

    /// The Universal Real Time SysEx that locates a receiver to `timecode`.
    ///
    /// `F0 7F <device> 01 01 hh mm ss ff F7`, addressed to `0x7F` (all devices).
    /// The hour byte packs the rate bits above the hour: `(rateBits << 5) | hh`.
    static func fullFrameBytes(_ timecode: Timecode) -> [UInt8] {
        let hourByte = (rateBits(for: timecode.rate) << 5) | (UInt8(timecode.hours) & 0x1F)
        return [
            0xF0, 0x7F, 0x7F,   // SysEx start, Universal Real Time, all devices
            0x01, 0x01,         // sub-ID 1: MIDI Time Code — sub-ID 2: Full Message
            hourByte,
            UInt8(timecode.minutes),
            UInt8(timecode.seconds),
            UInt8(timecode.frames),
            0xF7                // SysEx end
        ]
    }
}

File: windows/OnlyCue.Core/Midi/MtcFrame.cs
namespace OnlyCue.Core.Midi;

/// <summary>
/// The MIDI Timecode wire format — pure byte encoding, no MIDI transport.
/// Mirrors the Swift <c>MTCFrame</c> (<c>OnlyCue/MIDI/MTCFrame.swift</c>); macOS
/// is the source of truth and <c>golden/mtc-wire-v1.json</c> is the contract.
/// </summary>
/// <remarks>
/// <para>
/// Two message shapes carry MTC: <b>quarter-frame</b> (<c>F1 &lt;data&gt;</c>),
/// the running stream, where a complete value takes eight messages and therefore
/// spans two frames; and <b>Full Frame</b>
/// (<c>F0 7F 7F 01 01 hh mm ss ff F7</c>), a Universal Real Time SysEx that jams
/// a receiver to an absolute position in one message.
/// </para>
/// <para>
/// <b>Rate bits are two bits wide</b> — MTC encodes exactly 24 / 25 / 29.97-DF /
/// 30 and has no 30-fps-drop-frame code, so <c>Fps30Drop</c> takes the 29.97-DF
/// slot while still being clocked at 30 fps (ADR-019, reaffirmed by ADR-032).
/// </para>
/// <para>
/// Every shift below is cast back to <c>byte</c> explicitly: in C# the operands
/// of <c>&lt;&lt;</c> and <c>&gt;&gt;</c> promote to <c>int</c>, so an
/// unintended width change is one missing cast away.
/// </para>
/// </remarks>
public static class MtcFrame
{
    /// <summary>Status byte introducing a quarter-frame message.</summary>
    public const byte QuarterFrameStatus = 0xF1;

    /// <summary>Quarter-frame messages needed to transmit one complete timecode.</summary>
    public const int PiecesPerTimecode = 8;

    /// <summary>Frames spanned by one full eight-message quarter-frame sequence.</summary>
    public const int FramesPerSequence = 2;

    /// <summary>The two-bit MTC rate code for <paramref name="rate"/>.</summary>
    public static byte RateBits(SmpteFramerate rate) => rate switch
    {
        SmpteFramerate.Fps24 => 0b00,
        SmpteFramerate.Fps25 => 0b01,
        SmpteFramerate.Fps30Drop => 0b10,   // 29.97 drop — the only drop-frame slot MTC has
        SmpteFramerate.Fps30 => 0b11,
        _ => throw new ArgumentOutOfRangeException(nameof(rate))
    };

    /// <summary>
    /// The data byte for quarter-frame <paramref name="piece"/> (0…7) of
    /// <paramref name="timecode"/>. Layout is <c>0nnn dddd</c>: <c>nnn</c> is the
    /// piece index, <c>dddd</c> the payload nibble. Pieces run
    /// least-significant-first — frames, seconds, minutes, hours — with the rate
    /// bits riding in piece 7 alongside the single high bit of the hour.
    /// </summary>
    /// <remarks><paramref name="piece"/> is <b>clamped</b> to 0…7 rather than
    /// rejected: an out-of-range index is a caller bug, and clamping keeps a
    /// malformed byte (one whose high nibble names a different piece) off the
    /// wire. Throwing here instead would be a silent divergence from macOS, so
    /// the golden vectors pin <c>-1</c> and <c>8</c> explicitly.</remarks>
    public static byte QuarterFrameByte(int piece, Timecode timecode)
    {
        var index = Math.Clamp(piece, 0, PiecesPerTimecode - 1);
        var payload = index switch
        {
            0 => (byte)(timecode.Frames & 0x0F),
            1 => (byte)((timecode.Frames >> 4) & 0x01),
            2 => (byte)(timecode.Seconds & 0x0F),
            3 => (byte)((timecode.Seconds >> 4) & 0x03),
            4 => (byte)(timecode.Minutes & 0x0F),
            5 => (byte)((timecode.Minutes >> 4) & 0x03),
            6 => (byte)(timecode.Hours & 0x0F),
            _ => (byte)((RateBits(timecode.Rate) << 1) | ((timecode.Hours >> 4) & 0x01))
        };

        return (byte)((index << 4) | payload);
    }

    /// <summary>The complete two-byte quarter-frame message.</summary>
    public static byte[] QuarterFrameMessage(int piece, Timecode timecode) =>
        [QuarterFrameStatus, QuarterFrameByte(piece, timecode)];

    /// <summary>
    /// The Universal Real Time SysEx that locates a receiver to
    /// <paramref name="timecode"/>: <c>F0 7F &lt;device&gt; 01 01 hh mm ss ff F7</c>,
    /// addressed to <c>0x7F</c> (all devices). The hour byte packs the rate bits
    /// above the hour: <c>(rateBits &lt;&lt; 5) | hh</c>.
    /// </summary>
    public static byte[] FullFrameBytes(Timecode timecode)
    {
        var hourByte = (byte)((RateBits(timecode.Rate) << 5) | (timecode.Hours & 0x1F));
        return
        [
            0xF0, 0x7F, 0x7F,   // SysEx start, Universal Real Time, all devices
            0x01, 0x01,         // sub-ID 1: MIDI Time Code — sub-ID 2: Full Message
            hourByte,
            (byte)timecode.Minutes,
            (byte)timecode.Seconds,
            (byte)timecode.Frames,
            0xF7                // SysEx end
        ];
    }
}
